### DISCUSSION TITLE
How to extract machine into component

### DISCUSSION THREAD

**Discussion by azoriusss** [chakra-ui/zag#1970] [Q&A] [answered]
Hi everyone !
I'm using zag.js for a project and it works pretty fine overall, but i noticed a weird behaviour when using sub-component in my machine

If I write this ( Disclaimer : this is a useless case juste for example)
``` 
export default function Checkbox() {
  const [state, send] = useMachine(
    checkbox.machine({
      id: useId(),
    })
  );

  const api = checkbox.connect(state, send, normalizeProps);

  const CheckboxContent = () => (
    <>
      <div {...api.getControlProps()} className="control" />

      <span {...api.getLabelProps()}>
        Input is {api.checked ? "checked" : "unchecked"}
      </span>
      <input {...api.getHiddenInputProps()} />
    </>
  );

  return (
    <label {...api.getRootProps()} className="root">
      <CheckboxContent />
    </label>
  );
}
```
Then the checkbox does not work. 
However, if I use `CheckboxContent()` instead of `<CheckboxContent/>`, it works fine
No problem either if I use the content of `CheckboxContent` directly

Is this  expected behaviour ? Is there some work around ? 

Thanks !

**Comment by segunadebayo** [comment, marked as answer]
You want to create a React context for subcomponents and share the `api` between them.

Here's an example
https://stackblitz.com/edit/8ur6n2?file=src%2FApp.tsx,src%2FCheckbox.tsx&showSidebar=0

**Reply by azoriusss** [reply]
Thank you for your answer !
So this behaviour is expected right ?
